File: dart/algorithms/bin/dynamic_programming.dart
class DynamicProgramming {
  var cache = <int, int>{};
  var count = 0;

  int fibonacci(int value) {
    if (value < 2) return value;
    if (cache.containsKey(value)) return cache[value]!;
    cache[value] = fibonacci(value - 1) + fibonacci(value - 2);
    ++count;
    return fibonacci(value - 1) + fibonacci(value - 2);
  }
}

File: kotlin/udemy_course/src/main/kotlin/recursion_backtracking_dp/DynamicProgramming.kt
package recursion_backtracking_dp

class DynamicProgramming {
    fun startDynamicProgramming() {
        println("Starting dynamic programming")

        println(fibonacciNumbers(5))
    }

    private fun fibonacciNumbers(n: Int, memorization: MutableMap<Int, Int> = mutableMapOf(0 to 1, 1 to 1)): Int {
        if (!memorization.containsKey(n)) {
            memorization[n] = fibonacciNumbers(n - 1, memorization) + fibonacciNumbers(n - 2, memorization)
        }
        return memorization[n]!!
    }
}